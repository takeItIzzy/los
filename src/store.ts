import createSubscribe, { StateBucket, Subscribe } from './utils/createSubscribe';

/**
 * Collecting state to store when useLosState or init/useLosInit is executing.
 * useLosValue triggers component re-rendering when state updated.
 */

export interface StoreItem {
  hasInit: boolean;
  value: any;
  reducer?: LosReducer<any, any>;
  stateBucket: StateBucket;
  subscribe: Subscribe;
}
export const store: Map<Atom<any, any>, StoreItem> = new Map();

export type LosAction<T> = { type: T; [key: string]: any };
export type LosReducer<S, A> = (state: S, action: LosAction<A>) => S;
export class Atom<T, A = void> {
  constructor(value?: T, reducer?: LosReducer<T, A>) {
    this.value = value;
    this.reducer = reducer;
  }
  value?: T;
  reducer?: LosReducer<T, A>;
}

export const atom = <T, A = void>(
  config: {
    defaultValue?: T;
    reducer?: LosReducer<T, A>;
  } = {}
): Atom<T, A> => {
  const { defaultValue, reducer } = config;
  const atomItem = new Atom(defaultValue, reducer);

  const stateBucket: StateBucket = new Set();
  store.set(atomItem, {
    hasInit: false,
    value: atomItem.value,
    reducer: atomItem.reducer,
    stateBucket,
    subscribe: createSubscribe(stateBucket),
  });

  return atomItem;
};
