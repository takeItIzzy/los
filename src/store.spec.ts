import { atom, LosReducer, store } from './store';

describe('atom() testing', () => {
  it('should exist current key in store when defaultValue is provided', () => {
    const state = atom({
      defaultValue: 'default',
    });
    expect(store.has(state)).toBeTruthy();
    expect(store.get(state)).toMatchSnapshot(`{
      hasInit: false,
      value: 'default',
      reducer: undefined,
      stateBucket: new Set(),
      subscribe: [Function]
    }`);
  });

  it('should exist current key in store when reducer is provided', () => {
    const reducer: LosReducer<any, any> = (state) => {
      return state;
    };
    const state = atom({
      reducer,
    });
    expect(store.has(state)).toBeTruthy();
    expect(store.get(state)).toMatchSnapshot(`{
      hasInit: false,
      value: undefined,
      reducer,
      stateBucket: new Set(),
      subscribe: [Function]
    }`);
  });

  it('should exist current key in store when defaultValue and reducer are both provided', () => {
    const reducer: LosReducer<any, any> = (state) => {
      return state;
    };
    const state = atom({
      defaultValue: 'default',
      reducer,
    });
    expect(store.has(state)).toBeTruthy();
    expect(store.get(state)).toMatchSnapshot(`{
      hasInit: false,
      value: 'default',
      reducer,
      stateBucket: new Set(),
      subscribe: [Function]
    }`);
  });

  it('should exist current key in store when defaultValue and reducer are neither provided', () => {
    const state = atom();
    expect(store.has(state)).toBeTruthy();
    expect(store.get(state)).toMatchSnapshot(`{
      hasInit: false,
      value: undefined,
      reducer: undefined,
      stateBucket: new Set(),
      subscribe: [Function]
    }`);
  });
});
