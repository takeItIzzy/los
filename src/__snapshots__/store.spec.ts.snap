// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atom() testing should exist current key in store when defaultValue and reducer are both provided: {
      hasInit: false,
      value: 'default',
      reducer,
      stateBucket: new Set(),
      subscribe: [Function]
    } 1`] = `
Object {
  "hasInit": false,
  "reducer": [Function],
  "stateBucket": Set {},
  "subscribe": [Function],
  "value": "default",
}
`;

exports[`atom() testing should exist current key in store when defaultValue and reducer are neither provided: {
      hasInit: false,
      value: undefined,
      reducer: undefined,
      stateBucket: new Set(),
      subscribe: [Function]
    } 1`] = `
Object {
  "hasInit": false,
  "reducer": undefined,
  "stateBucket": Set {},
  "subscribe": [Function],
  "value": undefined,
}
`;

exports[`atom() testing should exist current key in store when defaultValue is provided: {
      hasInit: false,
      value: 'default',
      reducer: undefined,
      stateBucket: new Set(),
      subscribe: [Function]
    } 1`] = `
Object {
  "hasInit": false,
  "reducer": undefined,
  "stateBucket": Set {},
  "subscribe": [Function],
  "value": "default",
}
`;

exports[`atom() testing should exist current key in store when reducer is provided: {
      hasInit: false,
      value: undefined,
      reducer,
      stateBucket: new Set(),
      subscribe: [Function]
    } 1`] = `
Object {
  "hasInit": false,
  "reducer": [Function],
  "stateBucket": Set {},
  "subscribe": [Function],
  "value": undefined,
}
`;
